### PR TITLE
CLOUDSTACK-8415. SSVM shutdown during snapshot operation leaves behin…

### PR DESCRIPTION
**Pull request by likitha**
…d partial unusable disks and VM snapshots in secondary storage.

Enhance storage garbage collector to delete corresponding partial disks created in secondary storage and VM snapshots while expunging 'ERROR' snapshot.

### CODE REVIEW

**Comment by wilderrodrigues** [comment]
LGTM :+1:

File: engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java [hunk 1]
```diff
@@ -172,7 +175,11 @@ public DataObject create(DataObject obj, DataStore dataStore) {
                     if (snapshotDataStoreVO != null) {
                         ss.setParentSnapshotId(snapshotDataStoreVO.getSnapshotId());
                     }
-                    ss.setInstallPath(TemplateConstants.DEFAULT_SNAPSHOT_ROOT_DIR + "/" + snapshotDao.findById(obj.getId()).getAccountId() + "/" + snapshot.getVolumeId());
+                    String snapshotInstallPath = TemplateConstants.DEFAULT_SNAPSHOT_ROOT_DIR + "/" + snapshotDao.findById(obj.getId()).getAccountId() + "/" + snapshot.getVolumeId();
+                    if (snapshot.getHypervisorType().equals(HypervisorType.VMware)) {
```
Comment (DaanHoogland) [inline engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java]: Is this really only true for vmware? It seems that adding a uuid to the path is a rather universal action in the ssvm, no?

File: server/src/com/cloud/storage/StorageManagerImpl.java [hunk 2]
```diff
@@ -1120,14 +1123,60 @@ public void cleanupStorage(boolean recurring) {
...
+                    List<SnapshotVO> snapshots = _snapshotDao.listAllByStatusIncludingRemoved(Snapshot.State.Error);
                     for (SnapshotVO snapshotVO : snapshots) {
                         try {
                             List<SnapshotDataStoreVO> storeRefs = _snapshotStoreDao.findBySnapshotId(snapshotVO.getId());
+                            boolean isVMware = snapshotVO.getHypervisorType().equals(HypervisorType.VMware);
+                            boolean removeSnapshot = true;
                             for (SnapshotDataStoreVO ref : storeRefs) {
-                                _snapshotStoreDao.expunge(ref.getId());
+                                // Cleanup corresponding items (if any) from secondary storage.
+                                if (isVMware) {
```
Comment (DaanHoogland) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: only for vmware???

**Comment by DaanHoogland** [comment]
This seems to me to be a solution for vmware for a more generic problem. in the generic code conditions for hypervisortype == vmware are set while the problem is generic over hypervisor typos.

File: server/src/com/cloud/storage/StorageManagerImpl.java (hunk 2 above)
Comment (wilderrodrigues) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: Good catch!

We would have to get the PR in and test it, without the IF, against Xen and KVM, at least. What do you think?

File: engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java (hunk 1 above)
Comment (wilderrodrigues) [inline engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java]: It seems like it should be applied to any hypervisor.

Is that another one we can take in and fix?

File: engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java [hunk 3]
```diff
@@ -172,7 +175,11 @@ public DataObject create(DataObject obj, DataStore dataStore) {
...
+                        snapshotInstallPath = snapshotInstallPath.concat("/" + UUID.randomUUID().toString());
```
Comment (wilderrodrigues) [inline engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java]: I would also call contact twice: 1 for the "/"; and the 2nd time for the UUID.randomUUID().toString().

Comment (DaanHoogland) [inline engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java]: auto spell typo? contact == concat

Comment (wilderrodrigues) [inline engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java]: Yep... I edited. :)

**Comment by remibergsma** [comment]
@likitha Any updates?

**Comment by remibergsma** [comment]
Who wants to step in and finish this work? It seems the original author is not able to finish it. If no one steps in, we'll have to close the PR without merging it so please help :-).

**Comment by rhtyd** [comment]
tag:vmware-pickup

**Comment by jburwell** [comment]
@likitha is this PR still pertinent?

**Comment by rhtyd** [comment]
@jburwell Likitha has moved on, unlikely we'll hear from her or see further development activity

**Comment by DaanHoogland** [comment]
This is open to long with no progress

**Comment by rhtyd** [comment]
@DaanHoogland while this is open with no progress, can you check if the issue has been fixed or there exists a PR that works towards fixing the same. Let's not close old PRs that might still be useful.

**Comment by sureshanaparti** [comment]
Continued with the code changes here in a new PR #2090.

**Pull request closed without merging**